neps/optimizers/models/deepGP.py

```diff
@@ -0,0 +1,572 @@
+        total_scaled_loss = 0.0
+        for batch_idx, start_index in enumerate(range(0, n_examples_batch, batch_size)):
+            end_index = min(start_index + batch_size, n_examples_batch)
+            minibatch_size = end_index - start_index + 1
+
+            # if only one example in the batch, skip the batch.
+            # Otherwise, the code will fail because of batchnorm
+            if minibatch_size <= 1:
+                continue
+
+            indices = permutation[start_index:end_index]
+
+            batch_x, batch_budget, batch_lc, batch_y = (
+                x_train[indices],
+                train_budgets[indices],
+                learning_curves[indices],
+                y_train[indices],
+            )
+
+            # Zero backprop gradients
+            optimizer.zero_grad()
+
+            projected_x = nn(batch_x, batch_budget, batch_lc)
+            model.set_train_data(projected_x, batch_y, strict=False)
+            output = model(projected_x)
+
+            # Calc loss and backprop derivatives
+            loss = -mll(output, model.train_targets)  # type: ignore
+            episodic_loss_value: float = loss.detach().to("cpu").item()
+            # weighted sum over losses in the batch
+            total_scaled_loss = total_scaled_loss + episodic_loss_value * minibatch_size
+
+            mse = gpytorch.metrics.mean_squared_error(output, model.train_targets)
+            logger.debug(
+                f"Epoch {epoch_nr}  Batch {batch_idx} - MSE {mse:.5f}, "
+                f"Loss: {episodic_loss_value:.3f}, "
+                f"lengthscale: {model.covar_module.base_kernel.lengthscale.item():.3f}, "
+                f"noise: {model.likelihood.noise.item():.3f}, "  # type: ignore
+            )
+
+            loss.backward()
+            optimizer.step()
+
+        # Get average weighted loss over every batch
+        average_loss = total_scaled_loss / n_examples_batch
+        if average_loss < min_avg_loss_val:
+            min_avg_loss_val = average_loss
+            count_down = patience
+        elif early_stopping:
+            logger.debug(
+                f"No improvement over the minimum loss value of {min_avg_loss_val} "
+                f"for the past {patience - count_down} epochs "
+                f"the training will stop in {count_down} epochs"
+            )
+            count_down -= 1
+
+
+@dataclass
+class DeepGP:
+    """Gaussian process with a deep kernel."""
+
+    # Required
+    pipeline_space: SearchSpace
+
+    # Optional
+    learning_curve_pad_value: float = 0.0
+    root_directory: Path | None = None
+    # IMPORTANT: Checkpointing does not use file locking
+    # IMPORTANT: hence it is not suitable for multiprocessing settings
+    checkpoint_file: Path | str = "surrogate_checkpoint.pth"
+    checkpointing: bool = False
+    early_stopping: bool = True
+    batch_size: int = 64
+    n_epochs: int = 1000
+    patience: int = 10
+    refine_epochs: int = 50
+    perf_patience: int = 10
```

Comment:
And the best default value for the `perf_patience` is `1.2 * max_fidelity`, [here](https://github.com/automl/neps/blob/d6e123faf41dde2461241a20b2014e4345d9a2a3/neps/optimizers/bayesian_optimization/models/deepGP.py#L213)